### ISSUE TITLE
pg_dump/pg_restore a table failed as the table has been altered to use an external table as a leaf child partition

### ISSUE THREAD

**Issue by fanfuxiaoran** [greenplum-db/gpdb#3569]
1. create table and alter partition with an external table

```
createdb test
psql test
CREATE TABLE test (a int) 
DISTRIBUTED BY (a)
PARTITION BY RANGE (a)
(start(1) end(3) every(1));

CREATE EXTERNAL WEB TABLE exchagne_external_table (like test_1_prt_1)
EXECUTE 'echo -e "2\n1"' on host
FORMAT 'csv';

ALTER TABLE test
EXCHANGE PARTITION FOR (integer '2') WITH TABLE exchagne_external_table WITHOUT VALIDATION;

createdb newdb

```

2.pg_dump the database test
```
pg_dump -Fc test > test.dump

``` 
results:
```
pg_dump: NOTICE:  COPY ignores external partition(s)
```

3.pg_restore to newdb
```
pg_restore -d newdb db.dump
```  
results:
```
pg_restore: [archiver (db)] Error while PROCESSING TOC:
pg_restore: [archiver (db)] Error from TOC entry 233; 1259 16808 EXTERNAL TABLE exchange_partition_1_prt_2_external_partition__ gpadmin
pg_restore: [archiver (db)] could not execute query: ERROR:  relation "exchange_partition_1_prt_2_external_partition__" already exists
    Command was: CREATE EXTERNAL WEB TABLE exchange_partition_1_prt_2_external_partition__ (
    a integer,
    b integer
) EXECUTE E'echo -e...
pg_restore: [archiver (db)] Error from TOC entry 231; 1259 16761 TABLE exchange_partition gpadmin
pg_restore: [archiver (db)] could not execute query: ERROR:  zero-length delimited identifier at or near """"
LINE 9: ALTER TABLE exchange_partition EXCHANGE PARTITION "" WITH TA...
                                                          ^
    Command was: CREATE TABLE exchange_partition (
    a integer,
    b integer
) DISTRIBUTED BY (a) PARTITION BY RANGE(a)
          (
     ...
pg_restore: [archiver (db)] could not execute query: ERROR:  relation "public.exchange_partition" does not exist
    Command was: ALTER TABLE public.exchange_partition OWNER TO gpadmin;


pg_restore: [archiver (db)] Error from TOC entry 2165; 0 16761 TABLE DATA exchange_partition gpadmin
pg_restore: [archiver (db)] could not execute query: ERROR:  relation "exchange_partition" does not exist
    Command was: COPY exchange_partition (a, b) FROM stdin;

WARNING: errors ignored on restore: 4
```

**Comment by danielgustafsson** [comment]
> On 19 Oct 2017, at 05:43, Xiaoran Wang <notifications@github.com> wrote:
> 
> 	• create table and alter partition with an external table

This is a hard problem to work around, I managed to hack my way around it in #1940 for supporting this very case in pg_upgrade but in the end we decided it wasn’t worth the added complexity. pg_upgrade instead throw an error on the presence of such a hierarchy.

Fixing this for pg_dump/pg_restore doesn’t require the Oid synchronization that pg_upgrade faced, so ideally it should be less involved (perhaps parts of my patch in #1940 can be extracted to handle it?).
